### PR TITLE
WIP: Remove FIXME for planner's SplitUpdate feature

### PR DESCRIPTION
**Pull request by l-wang**
This fixs the following TODOs left in 73801e8d
1) Correct cost calculation for SplitUpdate plan. This will not help generating
a better plan because we have no choice other than simply adding the
SplitUpdate node.
2) Only send distribution columns for delete execution instead of all columns.

Co-authored-by: Shujie Zhang <shzhang@pivotal.io>
Co-authored-by: Alexandra Wang <leiwangcheme@gmail.com>
Signed-off-by: Alexandra Wang <leiwangcheme@gmail.com>

### CODE REVIEW

File: src/backend/cdb/cdbmutate.c [hunk 1]
```diff
@@ -1492,20 +1493,41 @@ correct_delete_idxes(List *deleteColIdx, List *targetList, List *varsAbsent, int
...
+	resultDesc = RelationGetDescr(resultRelation);
+	policy = resultRelation->rd_cdbpolicy;
 
 	for (attrIdx = 1; attrIdx <= resultDesc->natts; ++attrIdx)
 	{
 		TargetEntry			*tle;
 		Var					*splitVar;
 		TargetEntry			*splitTargetEntry;
 		Form_pg_attribute	attr;
+		int i = 0;
```
Comment (macroyuyang) [inline src/backend/cdb/cdbmutate.c]: indention here?

File: src/backend/cdb/cdbmutate.c [hunk 2]
```diff
@@ -1492,20 +1493,41 @@ correct_delete_idxes(List *deleteColIdx, List *targetList, List *varsAbsent, int
...
 
 		*insertColIdx = lappend_int(*insertColIdx, attrIdx);
-		*deleteColIdx = lappend_int(*deleteColIdx, attrIdx);
+		isPolicyKey = false;
+
+		/*
+		 * We only need the attrIdx of distribution keys for deletion,
+		 * this way we consume less network bandwidth during motion
+		 */
+		for (i = 0; i < policy->nattrs; i++)
```
Comment (macroyuyang) [inline src/backend/cdb/cdbmutate.c]: Do we need to consider partition table? For example, tbl (a, b) distributed by (a) partition by (b), if we only send column 'a' for deletion, the lookup for partition table will fail?

File: src/backend/cdb/cdbmutate.c [hunk 3]
```diff
@@ -1534,9 +1556,9 @@ process_targetlist_for_splitupdate(TupleDesc resultDesc, Index resultRelationsId
 		 *  so we record it as absent Vars, and we will add it to lower plan node in
 		 *  ensuing steps.
 		 */
-		if (IsA(tle->expr, Var) &&
+		if ((IsA(tle->expr, Var) &&
 			((Var *) tle->expr)->varnoold == resultRelationsIdx &&
-			((Var *) tle->expr)->varoattno == attrIdx)
+			((Var *) tle->expr)->varoattno == attrIdx) || !isPolicyKey)
```
Comment (kainwen) [inline src/backend/cdb/cdbmutate.c]: Should change a line.

**Comment by l-wang** [comment]
Closing this PR and reopening a new one here: https://github.com/greenplum-db/gpdb/pull/5698

The new PR only corrects the cost calculation. The optimization of sending less attributes for redistribution is a bit trivial and may not worth the effort at the moment.

**Pull request closed without merging**